core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java

```diff
@@ -4083,6 +4098,11 @@ private OperatorFactory createHashAggregationOperatorFactory(
         }
     }
 
+    private int getMaxWritersBasedOnMemory(Session session)
+    {
+        return (int) ceil((double) getQueryMaxMemoryPerNode(session).toBytes() / MAX_MEMORY_PER_PARTITION_WRITER);
```

Comment:
now that whole mechanism depends on `MAX_MEMORY_PER_PARTITION_WRITER`, I think it should be exposed as feature config/session prop